### ISSUE TITLE
[Bug] Memory overload during test sweeps

### ISSUE THREAD

**Issue by vbrkicTT** [tenstorrent/tt-forge-fe#867]
# Description

During running test sweeps memory usage constantly increase until pytest process is killed. Last reported memory usage is ~20GB.

Memory usage increment is usually small ~0.01 MB but occasionally higher ~50MB. During running sweeps tests where number of tests is high (>10 000 tests) memory usage accumulate until memory is overloaded and pytest process is killed.

| Number of tests | Memory usage |
| --- | --- |
| 500 | 3 GB |
| 2000 | 4.5 GB |
| 5000 | 10 GB |

Example of github action is https://github.com/tenstorrent/tt-forge-fe/actions/runs/12179328989/job/33971261971.

Pytest log

```
2024-12-05T12:40:29.4770087Z 2024-12-05 12:40:29.476 | INFO     | conftest:memory_usage_tracker:83 - Test memory usage:
2024-12-05T12:40:29.4770912Z 2024-12-05 12:40:29.476 | INFO     | conftest:memory_usage_tracker:84 -     By test: 269.52 MB
2024-12-05T12:40:29.4771689Z 2024-12-05 12:40:29.476 | INFO     | conftest:memory_usage_tracker:85 -     Minimum: 18766.89 MB
2024-12-05T12:40:29.4772506Z 2024-12-05 12:40:29.476 | INFO     | conftest:memory_usage_tracker:86 -     Maximum: 21736.70 MB
2024-12-05T12:40:29.4773311Z 2024-12-05 12:40:29.476 | INFO     | conftest:memory_usage_tracker:87 -     Average: 20065.86 MB
2024-12-05T12:40:29.4778366Z 
2024-12-05T12:40:35.2613401Z forge/test/operators/pytorch/test_all.py::test_plan[no_device-concatenate-CONST_EVAL_PASS-{'dim': 0}-(12, 64, 160, 96)-7-None-None] 2024-12-05 12:40:35.260 | INFO     | forge.compile:compile_main:219 - Compiling module ModelConstEvalPass
2024-12-05T12:40:35.2615619Z 2024-12-05 12:40:35.260 | INFO     | forge.compile:forge_compile_from_context:280 - Running compile stage init_compile
2024-12-05T12:40:35.2622185Z 2024-12-05 12:40:35.261 | INFO     | forge.ci:initialize_output_build_directory:100 - Forge output build directory for compiled artifacts: /tmp/root/3736448be14a
2024-12-05T12:40:35.2624727Z 2024-12-05 12:40:35.262 | INFO     | forge.ci:create_symlink:90 - Symlink created from /__w/tt-forge-fe/tt-forge-fe/tt_build/test_out to /tmp/root/3736448be14a
2024-12-05T12:40:35.2626372Z 2024-12-05 12:40:35.262 | INFO     | forge.compile:forge_compile_from_context:280 - Running compile stage generate_initial_graph
2024-12-05T12:40:43.0701198Z /__w/_temp/1bfebbc6-4766-4275-ba81-4b3cb5ef63a1.sh: line 10: 13846 Killed                  pytest -m push --splits 2 --group 1 --splitting-algorithm least_duration -m "nightly_sweeps" --junit-xml=reports/report_33971261971.xml 2>&1
2024-12-05T12:40:43.0702558Z      13847 Done                    | tee pytest.log
2024-12-05T12:40:43.1438119Z ##[error]Process completed with exit code 137.
```

Relates to PR #822

# Reproduce

```
# Checkout branch with added information about end memory after each tests (optional step)
git checkout vbrkic/skip-fatal-tests

# Run on 0.5% of tests
SAMPLE=0.5 pytest forge/test/operators/pytorch/test_all.py::test_query 2>&1 | grep 'test_all.py::test_query\|End: \|Average: '

# Some operators are more greedy than others
SAMPLE=0.5 OPERATORS=concatenate,Embedding pytest forge/test/operators/pytorch/test_all.py::test_query 2>&1 | grep 'test_all.py::test_query\|End: \|Average: '
```

**Comment by vbrkicTT** [comment]
@vmilosevic @nvukobratTT @dgolubovicTT 

It looks like we hit the memory issue again. Here is an example of a failing job https://github.com/tenstorrent/tt-forge-fe/actions/runs/13719977784/job/38374024420. Job is getting killed during execution.

**Comment by vbrkicTT** [comment]
Some experiments like calling garbage collector (don't resolve issue) are in #1400

@vmilosevic Is there a way to run tests in smaller chunks via pytest-split? Currently we have 30k tests split in 4 workers. If we could split pytest calls on each workers in smaller chunks with ~2k tests per pytest run we would avoid accumulation of memory above memory limits.

**Comment by vmilosevic** [comment]
Well we can adjust the number of parallel test runs to split into more groups, or we can call pytest multiple times in sequence.

Let's say for example we want tests to run on 4 runners in parallel, and to on each runner we want to split it into 10 separate pytest calls (10 subgroups).

```
runner 1:
--------
pytest --splits 40 --group 1
pytest --splits 40 --group 2
..
pytest --splits 40 --group 10

runner 2:
--------
pytest --splits 40 --group 11
pytest --splits 40 --group 12
..
pytest --splits 40 --group 20
...
```

```
for ((i=0; i<=${{ inputs.subgroups }}; i++)); do
    pytest -m push --splits ${{ inputs.test_group_cnt * inputs.subgroups }} \
        --group ${{ (matrix.test_group_id - 1)*inputs.subgroups + i }} \
        --splitting-algorithm least_duration \
        -m "${{ inputs.test_mark }}" \
        --junit-xml=${{ steps.strings.outputs.test_report_path }} \
        2>&1 | tee pytest.log
done
```

We should also check if collecting results works ok with multiple runs.

My concern with this is that we complicate test call, and we also hide some potential memory leaks and poor resource management.

**Comment by nvukobratTT** [comment]
Is this one still active?

**Comment by vbrkicTT** [comment]
> Is this one still active?

Situation is much better now.
Many memory management fixes has been merged recently that improved situation significantly.
We also need to skip few tests that require too much memory so after #1992 we expect no more crashing of github action.

However after adding new tests we might hit machine limits again in future, but for now it looks like we are fine.
